// ocPlatform/lib/mvc/view/View.js
var RenderBuffer = require("./RenderBuffer.js") ;
var Template = require("ocTemplate/lib/Template.js") ;


module.exports = function(tpl)
{
	this.tpl = tpl ;
	this._children = {} ;
	this._name = "" ;
	this.buff = new RenderBuffer ;
	this.wrapperClasses = ["ocview"] ;
	this.model = { $view: this } ;
}

module.exports.prototype.render = function(callback)
{
	if( this.tpl )
	{
		var view = this ;
		this.tpl.render( this.model, function(err){
			callback && callback(err,view) ;
		}, this.buff ) ;
	}
	else
	{
		callback && callback(null,this) ;
	}
}

module.exports.prototype.createWrapperHead = function(tagName)
{
	var html = "<"+ (tagName||'div') +' class="'+this.wrapperClasses.join(' ')+'"' ;
	html+= ' controller="'+(this.model.$controllerpath||"")+'"' ;
	html+= ' sumsign="'+(this.model.$sumsign||"...")+'"' ;
	if(this.tpl && this.tpl.onshowScript)
	{
		html+= ' onshow="'+this.tpl.onshowScript+'"' ;
	}
	html+= ">" ;

	return html ;
}

/**
 * 清理多余的对象引用，为输出到前端做准备
 */
module.exports.prototype.cleanup = function()
{
	this.tpl = this.tpl? this.tpl.filename: null ;
	if(this.buff)
	{
		this.buff.destroy() ;
	}
	this.buff = undefined ;
	this.queue = undefined ;
	this._children = undefined ;
	this._name = undefined ;
	this.model = undefined ;
}

module.exports.prototype.destroy = function()
{
	if(this.buff)
	{
		this.buff.destroy() ;
	}
	this.buff = undefined ;
	this.model = undefined ;
}



module.exports.__SHIPPABLE = true ;
