// ocPlatform/lib/mvc/Controller.js

if( typeof window=="undefined" )
{
	var path = require("path") ;
	var _url = require("url") ;
	var querystring = require("querystring") ;
}
var Class = require("ocClass/lib/Class.js") ;
var trace = require("stack-trace") ;
var View = require("./view/View.js") ;
var md5 = require("ocPlatform/lib/util/md5.js") ;
var utilarray = require("ocPlatform/lib/util/array.js") ;

var Controller = module.exports = Class.extend({

	title: ""
	, description: ""
	, keywords: []
	, view: null
	, viewReal: null
	, layout: null
	, children: {}
	, actions: {}

	, _initCallbacks: []
	, _initialized: 0
	, _initializeErr: null
	, init: function(callback){

		if(this._initialized==2)
		{
			callback(this._initializeErr,this) ;
			return ;
		}

		this._initCallbacks.push(callback) ;

		if(this._initialized==1)
		{
			return ;
		}

		this._initialized==1 ;


		var controller = this ;
		var waterfall = [
			this._loadView
			, this._initLayout
			, this._initChildren
			, this._initActions
		] ;

		var runstep = function ()
		{
			if(waterfall.length<=0)
			{
				// complete
				controller._emitInitCallbacks(null) ;
				return ;
			}

			var step = waterfall.shift() ;

			step.call( controller, function(err){
				if(err)
				{
					controller._emitInitCallbacks(err) ;
					return ;
				}

				// next step
				runstep() ;
			} );
		}

		// start
		runstep() ;
	}
	, _emitInitCallbacks: function(err){
		this._initialized = 2 ;
		this._initializeErr = err ;

		var callback ;
		while( callback=this._initCallbacks.shift() )
		{
			callback(this._initializeErr,this) ;
		}
	}

	, process: function(req,rsp){
		return true ;
	}

	, main: function(req,rsp,platform)
	{
		var processController = function(params,layout)
		{
			params = params || { req: req } ;
			var output = new Controller.Output(this,rsp) ;

			// 计算摘要签名
			this.evalSummarySignature(params,output) ;

			// 检查客户端缓存签名
			if( utilarray.search(sumsigns,output.model.$sumsign)===false )
			{
				// 执行自己
				rsp.wait() ;
				if( this.process(params,output,platform)===true )
				{
					rsp.end() ;
				}
			}

			// 执行 layout
			if( layout && this.layout )
			{
				output._children.layout = arguments.callee.call(this.layout,params.$layout,true) ;
				output._children.layout.view.wrapperClasses.push("oclayout") ;
			}

			// 执行子控制器
			for(var name in this.children)
			{
				output._children[name] = arguments.callee.call(this.children[name],params["$"+name],false) ;
			}

			return output ;
		}

		// top lock
		rsp.wait() ;


		var params = Controller.buildParamsFromRequest(req) ;
		var sumsigns = params["@sumsigns"]? params["@sumsigns"].splice(","): [] ;

		var output = processController.call(
				this
				, params
				, boolean(params,"@layout",true)
		) ;


		// render and output views
		if( boolean(params,"@render",true) )
		{
			rsp.wait() ;

			View.Caches.singleton.load("ocPlatform/templates/WebPage.html",null,function(err,webpage){

				if(err)
				{
					rsp.statusCode = "500" ;
					rsp.end() ;

					output.release() ;
					return ;
				}

				var ctx = webpage.createRenderContext() ;
				ctx._children["*"] = output.assembleView() ;
				ctx.render(function(err){

					rsp.setHeader("Content-Type","text/html") ;
					rsp.write( ctx.buff.toString() ) ;

					rsp.end() ;
					output.release() ;
				}) ;

			}) ;

		}
		else
		{
			output.cleanup() ;

			rsp.setHeader("Content-Type","application/javascript") ;
			rsp.write( JSON.stringify(output) ) ;
			rsp.end() ;

			output.release() ;
		}


		// top unlock
		rsp.end() ;
	}

	, child: function(name,includeAction){
		if( typeof this.children[name]!="undefined" )
		{
			return this.children[name] ;
		}
		else if(includeAction && typeof this.actions[name]!="undefined")
		{
			return this.actions[name] ;
		}
		else
		{
			return null ;
		}
	}

	, action: function(name,includeChild){
		if( typeof this.actions[name]!="undefined" )
		{
			return this.actions[name] ;
		}
		else if(includeChild && typeof this.children[name]!="undefined")
		{
			return this.children[name] ;
		}
		else
		{
			return null ;
		}
	}


	, _loadView: function(callback){

		if( !this.view )
		{
			callback(null,this) ;
			this.view = null ;
			return ;
		}

		if( typeof this.view!="string" )
		{
			callback(new Error("view property of Controller must be a string")) ;
			return ;
		}

		var viewrReal = View.resolve(this.view) ;
		var controller = this ;

		View.Caches.singleton.load(viewrReal,null,function(err,view){
			controller.view = view ;
			callback(err,controller) ;
		}) ;
	}

	, _initLayout: function(callback){

		if( !this.layout )
		{
			callback(null,this) ;
			this.layout = null ;
			return ;
		}

		var controller = this ;
		Controller.load(this.layout,function(err,layout){
			controller.layout = layout ;
			callback(err,controller) ;
		})
	}

	, _initChildren: function(callback,type){

		var type = type || "children" ;

		var childNames = [] ;
		for( var name in this[type] )
		{
			childNames.push(name) ;
		}

		var controller = this ;
		var loadChild = function()
		{
			if( childNames.length<=0 )
			{
				callback(null,controller) ;
				return ;
			}

			var name = childNames.shift() ;
			Controller.load(
				controller[type][name]
				, function(err,child){
					controller[type][name] = child ;
					if(err)
					{
						callback(err,controller) ;
					}
					else
					{
						loadChild() ;
					}
				}
				, typeof controller[type][name]!="string"?
						controller.constructor.pathname+":"+name :
						undefined
			)
		}
		loadChild() ;
	}

	, _initActions: function(callback){
		this._initChildren(callback,"actions") ;
	}

	, evalSummarySignature: function(params,output)
	{
		// 把params内容排序
		function sortParamsForSumsign(params)
		{
			var validnames = [] ;
			for(var name in params)
			{
				if(name[0]=="$"||name[0]=="@"||name=='req')	// '$','@'
				{
					continue ;
				}
				else if( typeof params[name]!="function" )
				{
					insertInOrder(validnames,name) ;
				}
			}
			var sorted = {} ;
			for(var i=0;i<validnames.length;i++)
			{
				sorted[validnames[i]] = typeof params[validnames[i]]=="object"?
							sortParamsForSumsign(params[validnames[i]]):
							params[validnames[i]] ;
			}
			return sorted ;
		}
		function insertInOrder(arr,name)
		{
			for(var i=arr.length-1;i>=0;i--)
			{
				if( arr[i] < name )
				{
					arr.splice(i+1,0,name) ;
					return ;
				}
			}
			arr.unshift(name) ;
		}

		var src = output.model.$controllerpath ;
		src+= JSON.stringify( sortParamsForSumsign(params) ) ;

		output.model.$sumsign = md5(src) ;
	}
}

// define static members, variable "this" is point to the class -------------
,{
	pathname: null

	, load: function(define,callback,pathname){

		callback = callback || function(){} ;

		if(typeof define=="string")
		{
			// 处理 action/child 格式
			var names = define.split(":") ;
			if(names.length>1)
			{
				return this.load(names[0],function(err,controller){
					if(err)
					{
						callback(err) ;
						return ;
					}

					for(var i=1;i<names.length;i++)
					{
						var controller = controller.action(names[i],true) ;
						if(!controller)
						{
							callback(new Error("cound not found controller by path: "+names.slice(0,i+1).join(":"))) ;
							return ;
						}
					}

					callback(null,controller) ;
				}) ;
			}

			try{
				var fullpath = Controller.resolve(define) || define ;
			}catch(e){
				callback(e) ;
				return ;
			}

			define = require(fullpath) ;

			if(!pathname)
			{
				pathname = path.relative(process.cwd()+"/node_modules",fullpath) ;
			}
		}

		var self = this ;
		var createControllerClass = function(module)
		{
			return (typeof module.Class=="undefined")? self.extend(module,{pathname:pathname}): module.Class ;
		}

		switch ( typeof define )
		{
			case "object" :
				define = define.Class = createControllerClass(define) ;
				break ;

			case "function" :
				if( !Class.isA(define,Controller) )
				{
					define = define.Class = createControllerClass({ process: define }) ;
				}
				break ;
			default :
				throw new Error("文件没有返回有效的控制器定义："+arguments[0]) ;
				break ;
		}

		return define.singleton(callback) ;
	}

	, singleton: function(callback){

		if(!this._singleton)
		{
			this._singleton = new this ;
			this._singleton.init(callback) ;
		}
		else
		{
			callback(null,this._singleton) ;
		}

		return this._singleton ;
	}

	, _singleton: null

	, extend: function(dynamicMembers,staticMembers){
		if( typeof dynamicMembers.config!=="undefined" )
		{
			// 和父类的 config 属性合并
			dynamicMembers.config = this.mergeObject(null,this.prototype.config,dynamicMembers.config) ;
		}

		return this._super(dynamicMembers,staticMembers) ;
	}

}) ;



// not inherists static members -------------

Controller._controllerAlias = {

	// layouts
	"webpage": "ocPlatform/lib/layout/WebPage.js"
	, "weblayout": "ocPlatform/lib/layout/WebLayout.js"
	, "controllpanel": "ocPlatform/lib/layout/LayoutControllPanel.js"
	, "desktop": "ocPlatform/layout/lib/LayoutDesktop.js"

	// controllers
	, "hello": "ocPlatform/lib/controllers/Hello.js"
}


Controller.resolve = function (path)
{
	if( path in Controller._controllerAlias )
	{
		path = Controller._controllerAlias[path] ;
	}

	try{
		return require.resolve(path) ;
	}catch(err){
		if( typeof err.code=="undefined" || err.code!="MODULE_NOT_FOUND" )
		{
			throw err ;
		}
	}

	// 清理开头的斜线 和 连续的斜线
	var subnames = path.replace(/^\/+/,"").replace(/\/+/g,"/").split("/") ;
	subnames.splice(1,0,"lib") ;
	path = subnames.join("/") ;

	return require.resolve(path) ;
}

Controller.buildParamsFromRequest = function(req)
{
	var url = _url.parse(req.url) ;
	var params = querystring.parse(url.query||"") ;

	return params ;
}


Controller.Output = function(controller,rsp)
{
	this.rsp = rsp ;
	this.messages = [] ;
	this.view = controller.view?
					controller.view.createRenderContext():
					View.createNullContext()
	this.model = this.view.model ;
	this.model.$controllerpath = controller.constructor.pathname ;

	this._children = {} ;
}

Controller.Output.prototype.assembleView = function()
{
	var rootView = this.view ;

	for(var name in this._children)
	{
		var child = this._children[name] ;


		if(name=="layout")
		{
			child.view._children["*"] = this.view ;

			// 递归
			rootView = child.assembleView() ;
		}
		else
		{
			this.view._children[name] = child.view ;

			// 递归
			child.assembleView() ;
		}
	}

	return rootView ;
}

/**
 * 清理多余的引用，为输出到前端做准备
 */
Controller.Output.prototype.cleanup = function()
{
	this.rsp = undefined ;
	this.view.cleanup() ;

	for(var name in this._children)
	{
		this._children[name].cleanup() ;
	}
}

/**
 *
 */
Controller.Output.prototype.release = function()
{
	this.rsp = null ;
	this.messages = null ;
	this.model = null ;
	this.view.release() ;

	for(var name in this._children)
	{
		this._children[name].release() ;
		this._children[name] = null ;
	}
	this._children = null ;
}

var FrontendViewTemplate = require("ocPlatform/lib/frontend/mvc/ViewTemplate.js") ;

/**
 * 从 json 还原回 Output对象
 * 在 frontend 调用
 */
Controller.Output.restore = function(output,callback)
{
	output.assembleView = Controller.Output.prototype.assembleView ;
	output.release = Controller.Output.prototype.release ;
	output.cleanup = Controller.Output.prototype.cleanup ;
	output.constructor = Controller.Output ;

	var childNames = [] ;
	for(var name in output._children)
	{
		childNames.push(name) ;
	}

	(function processChild(err,child){
		if(err)
		{
			callback(err,output) ;
		}

		// all children 已完成
		else if( !childNames.length )
		{
			// 加载自己的视图
			if(output.view.tpl)
			{
				FrontendViewTemplate.Caches.singleton.load(output.view.tpl,function(err,tpl){
					if(!err)
					{
						output.view = tpl.createRenderContext() ;
						output.view.model = output.model ;
					}
					callback(err,output) ;
				}) ;
			}
			else
			{
				callback(null,output) ;
			}
		}

		// process a child
		else
		{
			var name = childNames.shift() ;
			Controller.Output.restore(output._children[name],processChild) ;
		}
	}) () ;
}


var boolean = function(model,name,defval)
{
	if( typeof model[name]=='undefined' )
	{
		return defval ;
	}
	return !model[name].match(/^(0|false)$/i) ;
}
