// ocPlatform/lib/mvc/MessageQueue.js
var Class = require("ocClass") ;
var strutil = require("ocPlatform/lib/util/string.js") ;

var MessageQueue = module.exports = function()
{}
MessageQueue.prototype = [] ;

MessageQueue.prototype.createMessage = function(sentence,args,type,title)
{
	this.push(new MessageQueue.Message(sentence,args,type,title)) ;
}

MessageQueue.restore = function(lst)
{
	lst.__proto__ = MessageQueue.prototype ;

	for(var i=0;i<lst.length;i++)
	{
		lst[i].__proto__ = MessageQueue.Message.prototype ;
	}
}


// ------------------------------------------------

MessageQueue.Message = function(sentence,args,type,title)
{
	this.type = type || 'warning' ;
	this.sentence = sentence ;
	this.args = args || [] ;
	this.title = title ;
}

MessageQueue.Message.prototype.toString = function()
{
	return strutil.sprintf(this.sentence,this.args) ;
}

MessageQueue.Message.prototype.toBootstrap = function()
{
	var html = '<div class="alert alert-'+this.type+'">' ;
	html+= '<button type="button" class="close" data-dismiss="alert">&times;</button>' ;
	html+= '<strong>'+ (this.title || MessageQueue.Message.defaultTitles[ this.type ] || '') +'</strong> ' ;
	html+= this.toString() ;
	html+= '</div>' ;
	return html ;
}

MessageQueue.Message.types = {
	success: "success"
	, error: "error"
	, danger: "danger"
	, info: "info"
	, warning: "warning"
}
MessageQueue.Message.defaultTitles = {
	success: "Success!"
	, error: "Error!"
	, danger: "Danger!"
	, info: "Information!"
	, warning: "Warning!"
}
