// ocPlatform/lib/mvc/MessageQueue.js
var Class = require("ocClass") ;
var util = require("util") ;

var MessageQueue = module.exports = function()
{}
MessageQueue.prototype = [] ;

MessageQueue.prototype.createMessage = function(sentence,args,type,title)
{
	this.push(new MessageQueue.Message(type,sentence,args,title)) ;
}

MessageQueue.Message = function(sentence,args,type,title)
{
	this.type = type ;
	this.sentence = sentence ;
	this.args = args || [] ;
	this.title = title ;
}

MessageQueue.Message.prototype.toString = function()
{
	return util.format(this.sentence,this.args) ;
}

MessageQueue.Message.prototype.toBootstrap = function()
{
	var html = '<div class="alert">' ;
	html+= '<button type="button" class="close" data-dismiss="alert">&times;</button>' ;
	html+= '<strong>'+ this.title || MessageQueue.Message.defaultTitles[ this.type ] +'</strong>' ;
	html+= this.toString() ;
	html+= '</div>' ;
	return html ;
}


MessageQueue.Message.types = {
	success: "success"
	, error: "error"
	, danger: "danger"
	, info: "info"
}
MessageQueue.Message.defaultTitles = {
	success: "Success!"
	, error: "Error!"
	, danger: "Danger!"
	, info: "Information!"
}
