// ocPlatform/lib/mvc/MessageQueue.js
var Class = require("ocClass") ;
var strutil = require("ocPlatform/lib/util/string.js") ;
var View = require("./view/View.js") ;

var MessageQueue = module.exports = function()
{}
MessageQueue.prototype = [] ;

MessageQueue.prototype.createMessage = function(sentence,args,type,title)
{
	this.push(new MessageQueue.Message(sentence,args,type,title)) ;
}

MessageQueue.prototype.render = function(callback,tplname)
{
	var i = -1 ;
	var queue = this ;

	(function step(err,buff)
	{
		if( err || (++i)>=queue.length )
		{
			callback(err,queue) ;
			return ;
		}

		queue[i].render( arguments.callee, tplname ) ;
	}) () ;
}

MessageQueue.restore = function(lst)
{
	lst.__proto__ = MessageQueue.prototype ;

	for(var i=0;i<lst.length;i++)
	{
		lst[i].__proto__ = MessageQueue.Message.prototype ;
	}
}


// ------------------------------------------------

MessageQueue.Message = function(sentence,args,type,title)
{
	this.type = type || 'warning' ;
	this.sentence = sentence ;
	this.args = args || [] ;
	this.title = title ;
	this.buff = null ;
}

MessageQueue.Message.prototype.cleanup = function()
{
	this.buff = null ;
}

MessageQueue.Message.prototype.toString = function()
{
	return strutil.sprintf(this.sentence,this.args) ;
}

MessageQueue.Message.prototype.toBootstrap = function()
{
	var html = '<div class="alert alert-'+this.type+'">' ;
	html+= '<button type="button" class="close" data-dismiss="alert">&times;</button>' ;
	html+= '<strong>'+ (this.title || MessageQueue.Message.defaultTitles[ this.type ] || '') +'</strong> ' ;
	html+= this.toString() ;
	html+= '</div>' ;
	return html ;
}

MessageQueue.Message.prototype.render = function(callback,tplname){

	if( this.buff )
	{
		callback && callback(null,this.buff) ;
		return ;
	}

	var message = this ;

	View.Caches.singleton.load( tplname||MessageQueue.Message.defautlTplName, function(err,tpl){

		if(err)
		{
			callback && callback(err) ;
			return ;
		}

		tpl.createRenderContext().render(function(err,buff){

			callback && callback(err,buff) ;

		},{message:message}) ;

	} )
}

MessageQueue.Message.defautlTplName = "ocPlatform/templates/message.bootstrap.html" ;


MessageQueue.Message.types = {
	success: "success"
	, error: "error"
	, danger: "danger"
	, info: "info"
	, warning: "warning"
}
MessageQueue.Message.defaultTitles = {
	success: "Success!"
	, error: "Error!"
	, danger: "Danger!"
	, info: "Information!"
	, warning: "Warning!"
}
